src/ldap.js

```diff
@@ -0,0 +1,69 @@
+var url = require('url');
+var ldap = require('ldapjs');
+var _ = require('lodash');
+var debug = require('debug')('LDAPClient');
+
+class LDAPClient {
+  constructor(cfg) {
+    let tlsOptions = {
+      cert:   cfg.cert,
+      key:    cfg.key,
+    };
+    let port = url.parse(cfg.url).port;
+    if (port) {
+      tlsOptions.port = port;
+    }
+
+    this.client = ldap.createClient({
+      url: cfg.url,
+      tlsOptions,
+      timeout: 10 * 1000,
+      reconnect: true,
+    });
+  }
+
+  bind(user, password) {
+    debug(`bind(${user}, <password>)`);
+    return new Promise((accept, reject) => this.client.bind(
+      user, password, err => {
+      err ? reject(err) : accept();
+    }));
+  }
+
+  search(base, options) {
+    debug(`search(${base}, ${JSON.stringify(options)})`);
+    return new Promise((accept, reject) => this.client.search(
+      base, options, (err, res) => {
+      err ? reject(err) : accept(res);
+    }));
```

Comment:
Since we don't do anything smart... you could make a generic implementation here that just adds all the `entry` objects from the search to a list and returns that...

That would simplify a lot... ie. in `addRolesForQuery` you would get a list of entries, and you just do `.map(..).filter(...)` on.

In obscure cases where the list is so large that we run out of memory we just crash.
I doubt that will happen.. I suspect LDAP was designed in a world with different memory requirements.